hometask-2/src/routes/users.ts

```diff
@@ -1,70 +1,81 @@
 import express, { Request, Response } from 'express';
 
-import { db } from '../db';
+import { getUserById, addUser, updateUser, deleteUser, getAllUsers, getUsersByLoginSubstring } from '../data-access/user-repository';
 import { validateSchema } from '../middleware/validations';
 import { newUserSchema, updateUserSchema } from '../schemas/users';
-import { User } from '../types';
+import { ERRORS, User } from '../types';
 
 const router = express.Router();
 
 router.route('/:id')
-    .get((req: Request, res: Response) => {
+    .get(async (req: Request, res: Response) => {
         try {
-            const user = db.getUserById(req.params.id);
+            const user = await getUserById(req.params.id);
+
+            if (!user) {
+                throw new Error(ERRORS.USER_NOT_FOUND);
+            }
 
             res.json(user);
         } catch (err: any) {
             res.status(400).json({ message: err.message });
         }
     })
-    .put(validateSchema(updateUserSchema), (req: Request, res: Response) => {
+    .put(validateSchema(updateUserSchema), async (req: Request, res: Response) => {
         try {
-            const updatedUser = db.updateUser(req.params.id, req.body);
+            const updatedUser = await updateUser(req.params.id, req.body);
+
+            if (!updatedUser) {
+                throw new Error(ERRORS.USER_NOT_FOUND);
+            }
 
             res.json(updatedUser);
         } catch (err: any) {
             res.status(400).json({ message: err.message });
         }
     })
-    .delete((req: Request, res: Response) => {
+    .delete(async (req: Request, res: Response) => {
         try {
-            const deletedUserId = db.deleteUser(req.params.id);
+            const deletedUserId = await deleteUser(req.params.id);
 
-            res.json({ id: deletedUserId.id });
+            if (!deletedUserId) {
+                throw new Error(ERRORS.USER_NOT_FOUND);
+            }
+
+            res.json({ id: deletedUserId });
         } catch (err: any) {
             res.status(400).json({ message: err.message });
         }
-    })
+    });
 
 router.route('/')
-    .get((req: Request, res: Response) => {
+    .get(async (req: Request, res: Response) => {
         try {
             const loginSubstring = req.query.loginSubstring as string;
             const limit = parseInt(req.query.limit as string, 10) as number;
 
             let users: User[];
 
             if (typeof loginSubstring === 'string') {
-                users = db.getUsersByLoginSubstring(loginSubstring, limit);
+                users = await getUsersByLoginSubstring(loginSubstring, limit) as unknown as User[];
```

Comment:
This type of double casting to `unknown` the to something else is not good because it let's you convert any type into anything. So basically same thing as just casting to `any` or having no types at all.

One rule you can have for yourself is: Every time you fell the need to do this, means there's something wrong with types somewhere else.

I'm not sure what the problem is, but my guess is that it is happening because your normaliser function will return `User | null` so that null is making your life harder at this point. And the funny thing is that you will most likely not have any nulls here. Probably the only place where you might have null is when getting a single entity.

My advice is: remove the null cases from the normalisers, and deal with nulls as a separate step, that way the type here will be User[] instead of (User | null)[]

Reply:
Thanks! Fixed